### PR TITLE
CPU optimizations

### PR DESCRIPTION
**Pull request by SmirnovEgorRu**
This PR is changes from https://github.com/dmlc/xgboost/pull/4278 without pre-processing optimizations (it has been merged yet https://github.com/dmlc/xgboost/pull/4310)

In addition I refactored the code to make it simple. In previews version I used omp tasks, but they are not supported in MSVS by default, so I replaced them by simple omp parallel for.

Code changes looks massive, but it is hard to split them on small parts.

### CODE REVIEW

**Comment by trivialfis** [comment]
It seems this is based on a old commit.  Is it possible to make a clean rebase?

**Comment by hcho3** [comment]
I'll review this pull request after the 0.90 release

**Comment by hcho3** [comment]
@SmirnovEgorRu Looks like this pull request is breaking some distributed training tests: https://xgboost-ci.net/blue/organizations/jenkins/xgboost/detail/PR-4433/7/pipeline#step-145-log-581

**Comment by SmirnovEgorRu** [comment]
@hcho3 Yes, now I want to understand how distributed mode works. Do we have some materials about this?
Potentially, my changes should optimize distributed mode too, but need to change code.

**Comment by hcho3** [comment]
@SmirnovEgorRu The XGBoost paper have some details about distributed training. Roughly speaking, every occurrence of AllReduce causes synchronization of all nodes

**Comment by hcho3** [comment]
Let me look at this over the weekend

**Comment by trivialfis** [comment]
Hi, @SmirnovEgorRu @hcho3 @RAMitchell  I want to revert part of #4310 .  That PR improved performance by allocating buffer, specifically:
https://github.com/dmlc/xgboost/blob/8ddd2715eefa21555c77bbf1ac2ea314ac9f279e/src/common/hist_util.cc#L111

This makes it very hard to load sparse dataset. For example the *first line* of URL dataset from  #2326 has largest index `3224681`.  Which means my machine won't even be able load a single line of data.  And just for future reference, allocating memory based on feature index might not be a good idea.

**Comment by SmirnovEgorRu** [comment]
@hcho3 , looks like all issues are resolved, including distributed mode.

File: src/common/hist_util.h [hunk 1]
```diff
@@ -10,13 +10,54 @@
...
+
+ private:
+  T* ptr_ = nullptr;
+  bool do_free_ = false;
+  size_t required_size_;
+  T stack_mem_[MaxStackSize];
+};
+
+namespace tree {
+class SplitEvaluator;
```
Comment (trivialfis) [inline src/common/hist_util.h]: Is this used somewhere in this file?

File: src/common/hist_util.h [hunk 2]
```diff
@@ -201,62 +248,121 @@ class GHistIndexBlockMatrix {
...
+  inline void AddHistRow(bst_uint nid) {
+    if (data_arr_.size() <= nid) {
+      data_arr_.resize(nid + 1, nullptr);
     }
-    CHECK_EQ(row_ptr_[nid], kMax);
 
-    row_ptr_[nid] = data_.size();
-    data_.resize(data_.size() + nbins_);
+    if (data_arr_[nid] == nullptr) {
+      data_arr_[nid] = new std::vector<GradStatHist>;
```
Comment (trivialfis) [inline src/common/hist_util.h]: Why do you need to use new?  I would try to avoid raw new and delete all together.

File: src/common/hist_util.h [hunk 3]
```diff
@@ -266,21 +372,55 @@ class GHistBuilder {
...
-  void BuildBlockHist(const std::vector<GradientPair>& gpair,
-                      const RowSetCollection::Elem row_indices,
-                      const GHistIndexBlockMatrix& gmatb,
-                      GHistRow hist);
-  // construct a histogram via subtraction trick
-  void SubtractionTrick(GHistRow self, GHistRow sibling, GHistRow parent);
+  }
+
+void BuildBlockHist(const std::vector<GradientPair>& gpair,
+                                  const RowSetCollection::Elem row_indices,
```
Comment (trivialfis) [inline src/common/hist_util.h]: indention is incorrect.

File: tests/cpp/tree/test_quantile_hist.cc [hunk 4]
```diff
@@ -206,7 +214,8 @@ class QuantileHistMock : public QuantileHistMaker {
       }
 
       /* Now compare against result given by EvaluateSplit() */
-      RealImpl::EvaluateSplit(0, gmat, hist_, *(*dmat), tree);
+      EvaluateSplitsBatch(nodes, gmat, **dmat, hist_is_init, hist_buffers, const_cast<RegTree*>(&tree));
```
Comment (trivialfis) [inline tests/cpp/tree/test_quantile_hist.cc]: Why is a const_cast necessary?

File: src/tree/updater_quantile_hist.h [hunk 5]
```diff
@@ -15,54 +15,23 @@
 #include <vector>
 #include <string>
 #include <queue>
+#include <deque>
```
Comment (trivialfis) [inline src/tree/updater_quantile_hist.h]: I couldn't find where you use deque

File: src/common/hist_util.cc [hunk 6]
```diff
@@ -545,177 +544,135 @@ void GHistIndexBlockMatrix::Init(const GHistIndexMatrix& gmat,
...
-
-    const size_t istart = iblock*block_size;
-    const size_t iend = (((iblock+1)*block_size > nrows) ? nrows : istart + block_size);
+  if (iend < nrows - no_prefetch_size) {
     for (size_t i = istart; i < iend; ++i) {
-      const size_t icol_start = row_ptr[rid[i]];
-      const size_t icol_end = row_ptr[rid[i]+1];
+      const size_t icol_start = rid[i] * n_features;
+      const size_t icol_start_prefetch = rid[i+prefetch_offset] * n_features;
+      const size_t idx_gh = 2*rid[i];
```
Comment (trivialfis) [inline src/common/hist_util.cc]: Use some more verbose name like grad_hist, or gradient_histogram.

File: src/common/hist_util.cc [hunk 7]
```diff
@@ -545,177 +544,135 @@ void GHistIndexBlockMatrix::Init(const GHistIndexMatrix& gmat,
...
+      for (size_t j = icol_start_prefetch; j < icol_start_prefetch + n_features;
+          j += prefetch_step) {
+        PREFETCH_READ_T0(index + j);
       }
 
-      for (size_t j = icol_start; j < icol_end; ++j) {
-        const uint32_t idx_bin = 2*index[j];
-        const size_t idx_gh = 2*rid[i];
+      grad_stat.sum_grad += pgh[idx_gh];
+      grad_stat.sum_hess += pgh[idx_gh+1];
```
Comment (trivialfis) [inline src/common/hist_util.cc]: Please give some notes about memory layout.

File: src/common/hist_util.h [hunk 8]
```diff
@@ -201,62 +248,121 @@ class GHistIndexBlockMatrix {
...
- * \brief histogram of graident statistics for a single node.
- *  Consists of multiple GradStats, each entry showing total graident statistics
- *     for that particular bin
- *  Uses global bin id so as to represent all features simultaneously
+ * \brief used instead of GradStats to have float instead of double to reduce histograms
+ * this improves performance by 10-30% and memory consumption for histograms by 2x
+ * accuracy in both cases is the same
  */
-using GHistRow = Span<tree::GradStats>;
+struct GradStatHist {
```
Comment (trivialfis) [inline src/common/hist_util.h]: So can we remove `GradStats`?

File: src/tree/updater_quantile_hist.h [hunk 9]
```diff
@@ -134,42 +103,104 @@ class QuantileHistMaker: public TreeUpdater {
...
+      double time_init_new_node = 0;
+      double time_build_hist = 0;
+      double time_evaluate_split = 0;
+      double time_apply_split = 0;
+
+      inline void StartPerfMonitor() {
+        global_start = dmlc::GetTime();
+      }
+
+      inline void EndPerfMonitor() {
```
Comment (trivialfis) [inline src/tree/updater_quantile_hist.h]: This is from an old commit.  It's no longer here.  Could you make a clean rebase?

File: src/tree/updater_quantile_hist.cc [hunk 10]
```diff
@@ -184,49 +271,395 @@ void QuantileHistMaker::Builder::EvaluateSplits(
...
+    hist_.AddHistRow(nid);
+    if (sibling_nid > -1) {
+      hist_.AddHistRow(sibling_nid);
+    }
+    const size_t nrows = row_set_collection_[nid].Size();
+    const size_t n_local_blocks = nrows / block_size_rows + !!(nrows % block_size_rows);
+    const size_t n_local_histograms = std::min(nthread, n_local_blocks);
+
+    for (size_t j = 0; j < n_local_blocks; ++j) {
+      task_nid.push_back(nid);
```
Comment (trivialfis) [inline src/tree/updater_quantile_hist.cc]: Can this be done by one resize?

File: src/tree/updater_quantile_hist.cc [hunk 11]
```diff
@@ -184,49 +271,395 @@ void QuantileHistMaker::Builder::EvaluateSplits(
...
+  const size_t block_size_rows = 256;
+  const size_t nthread = static_cast<size_t>(this->nthread_);
+  const size_t nbins = gmat.cut.row_ptr.back();
+  const size_t hist_size = 2  * nbins;
+
+  hist_buffers->resize(nodes.size());
+  hist_is_init->resize(nodes.size());
+
+  // input data for tasks
+  int32_t n_tasks = 0;
```
Comment (trivialfis) [inline src/tree/updater_quantile_hist.cc]: What task?

**Pull request closed without merging**